arangod/IResearch/IResearchView.cpp

```diff
@@ -850,155 +741,143 @@ IResearchView::Snapshot const* IResearchView::snapshot(
           return ctx;  // ensure same collections
         }
-
         ctx->clear();  // reassemble snapshot
       }
       break;
     case SnapshotMode::SyncAndReplace: {
       if (ctx) {
         ctx->clear();  // ignore existing cookie, recreate snapshot
       }
-
       auto res = const_cast<IResearchView*>(this)->commit();
-
       if (!res.ok()) {
-        LOG_TOPIC("fd776", WARN, arangodb::iresearch::TOPIC)
+        LOG_TOPIC("fd776", WARN, TOPIC)
             << "failed to sync while creating snapshot for arangosearch view '"
             << name() << "', previous snapshot will be used instead, error: '"
             << res.errorMessage() << "'";
       }
-
-      break;
-    }
-    default:
+    } break;
+    default:              // TODO unreachable
       TRI_ASSERT(false);  // all values of the enum should be covered
   }
-
   if (!ctx) {
     auto ptr = irs::memory::make_unique<ViewTrxState>();
-
     ctx = ptr.get();
     state.cookie(key, std::move(ptr));
-
     if (!ctx) {
-      LOG_TOPIC("61271", WARN, arangodb::iresearch::TOPIC)
+      LOG_TOPIC("61271", WARN, TOPIC)
           << "failed to store state into a TransactionState for snapshot of "
              "arangosearch view '"
           << name() << "', tid '" << state.id() << "'";
-
       return nullptr;
     }
   }
-
-  read_write_mutex::read_mutex mutex(
-      _mutex);  // '_metaState' can be asynchronously modified
+  read_write_mutex::read_mutex mutex(_mutex);
+  // '_metaState' can be asynchronously modified
   auto lock = irs::make_lock_guard(mutex);
-
   try {
+    auto cantFind = [&] {
+      LOG_TOPIC("d63ff", ERR, arangodb::iresearch::TOPIC)
+          << "Some link was removed in arangosearch view '" << name()
+          << "', skipping it";
+      state.cookie(key, nullptr);  // unset cookie
+      return nullptr;              // skip missing links
+    };
+    auto shouldFind = collections->size();
+    auto maxFind = _links.size();
     // collect snapshots from all requested links
-    for (auto const cid : *collections) {
-      IResearchLink::Snapshot snapshot;
-
-      if (auto const itr = _links.find(cid);
-          itr != _links.end() && itr->second) {
-        auto link = itr->second->lock();
-
-        if (!link) {
-          LOG_TOPIC("d63ff", ERR, arangodb::iresearch::TOPIC)
-              << "failed to find an arangosearch link in collection '" << cid
-              << "' for arangosearch view '" << name() << "', skipping it";
-          state.cookie(key, nullptr);  // unset cookie
-
-          return nullptr;  // skip missing links
-        }
-
-        snapshot = link->snapshot();
+    for (auto& entry : _links) {
+      if (shouldFind > maxFind) {
+        return cantFind();
       }
-
+      --maxFind;
+      auto linkLock = entry.second->lock();
+      if (!linkLock) {
+        continue;
+      }
+      auto cid = linkLock->collection().id();
+      if (!collections->contains(cid)) {
+        continue;
+      }
```

Comment:
We can check `collections->containes` before `->lock()` if we would store `collection id` in `entry`.
But I think `shared_lock` is not critical and this code ok